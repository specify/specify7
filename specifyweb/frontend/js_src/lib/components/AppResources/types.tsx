import type { LocalizedString } from 'typesafe-i18n';

import { preferencesText } from '../../localization/preferences';
import { reportsText } from '../../localization/report';
import { resourcesText } from '../../localization/resources';
import type { IR, RA, RR } from '../../utils/types';
import { ensure } from '../../utils/types';
import { icons } from '../Atoms/Icons';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpAppResourceDir, Tables } from '../DataModel/types';
import type { AppResourceMode } from './helpers';

export type AppResourceScope =
  | 'collection'
  | 'discipline'
  | 'global'
  | 'user'
  | 'userType';

export type AppResourceType = {
  readonly tableName: keyof Tables & ('SpAppResource' | 'SpViewSetObj');
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
};

export const appResourceTypes: RR<AppResourceMode, AppResourceType> = {
  appResources: {
    tableName: 'SpAppResource',
    icon: icons.cog,
    label: resourcesText.appResource(),
  },
  viewSets: {
    tableName: 'SpViewSetObj',
    icon: icons.pencilAt,
    label: resourcesText.formDefinitions(),
  },
};

export type ScopedAppResourceDir = SerializedResource<SpAppResourceDir> & {
  readonly scope: AppResourceScope;
};

type AppResourceSubType = {
  readonly mimeType: string | undefined;
  readonly name: string | undefined;
  readonly documentationUrl: string | undefined;
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
  /**
   * Whether when creating a new app resource of this type, should copy the
   * contents from an existing app resource of that type that is in current
   * scope.
   * Default value:
   * If app resource type can only have one specific name, this is true
   * Else false
   */
  readonly useTemplate?: boolean;
  /**
   * Only allow creating this app resource at certain levels
   */
  readonly scope?: RA<AppResourceScope>;
};

/**
 * The order of the subtypes matters. The filtering algorithm loops over these
 * in the order they are defined to find the first subType that matches the
 * current resource. Thus, subtypes should be sorted from the most
 * specific to the least specific.
 */
export const appResourceSubTypes = ensure<IR<AppResourceSubType>>()({
  label: {
    mimeType: 'jrxml/label',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.ticket,
    label: reportsText.label(),
  },
  report: {
    mimeType: 'jrxml/report',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.documentReport,
    label: reportsText.report(),
  },
  userPreferences: {
    mimeType: 'application/json',
    name: 'UserPreferences',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/specify-7-user-preferences-webinar/861',
    icon: icons.cog,
    label: preferencesText.userPreferences(),
    useTemplate: false,
    scope: ['user'],
  },
  defaultUserPreferences: {
    mimeType: 'application/json',
    name: 'DefaultUserPreferences',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Setting-default-user-preferences',
    icon: icons.cog,
    label: preferencesText.defaultUserPreferences(),
  },
  collectionPreferences: {
    mimeType: 'application/json',
    name: 'CollectionPreferences',
    documentationUrl: undefined,
    icon: icons.cog,
    label: preferencesText.collectionPreferences(),
    scope: ['collection'],
  },
  leafletLayers: {
    mimeType: 'application/json',
    name: 'leaflet-layers',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Adding-Custom-Tile-Servers',
    icon: icons.locationMarker,
    label: resourcesText.leafletLayers(),
  },
  rssExportFeed: {
    mimeType: 'text/xml',
    name: 'ExportFeed',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Darwin-Core-Archive-Publishing',
    icon: icons.upload,
    label: resourcesText.rssExportFeed(),
  },
  expressSearchConfig: {
    mimeType: 'text/xml',
    name: 'ExpressSearchConfig',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/simple-search-config/183',
    icon: icons.search,
    label: resourcesText.expressSearchConfig(),
  },
  typeSearches: {
    mimeType: 'text/xml',
    name: 'TypeSearches',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/adding-a-non-native-query-combo-box/859#h-1-type-search-definition-typesearch_defxml-8',
    icon: icons.documentSearch,
    label: resourcesText.typeSearches(),
  },
  webLinks: {
    mimeType: 'text/xml',
    name: 'WebLinks',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/weblinks.xml',
    icon: icons.externalLink,
    label: resourcesText.webLinks(),
  },
  uiFormatters: {
    mimeType: 'text/xml',
    name: 'UIFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/uiformatters.xml',
    icon: icons.hashtag,
    label: resourcesText.uiFormatters(),
  },
  dataObjectFormatters: {
    mimeType: 'text/xml',
    name: 'DataObjFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/dataobj_formatters.xml',
    icon: icons.variable,
    label: resourcesText.dataObjectFormatters(),
  },
  dataEntryTables: {
    mimeType: 'text/xml',
    name: 'DataEntryTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/fish/dataentry_task.xml',
    icon: icons.pencilAt,
    label: resourcesText.dataEntryTables(),
  },
  interactionsTables: {
    mimeType: 'text/xml',
    name: 'InteractionsTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/interactionstask.xml',
    icon: icons.chat,
    label: resourcesText.interactionsTables(),
  },
  otherXmlResource: {
    mimeType: 'text/xml',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.code,
    label: resourcesText.otherXmlResource(),
  },
  otherJsonResource: {
    mimeType: 'application/json',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.adjustments,
    label: resourcesText.otherJsonResource(),
  },
  otherPropertiesResource: {
    mimeType: 'text/x-java-properties',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.viewList,
    label: resourcesText.otherPropertiesResource(),
  },
  otherAppResources: {
    mimeType: undefined,
    name: undefined,
    documentationUrl: undefined,
    icon: icons.document,
    label: resourcesText.otherAppResource(),
  },
} as const);
