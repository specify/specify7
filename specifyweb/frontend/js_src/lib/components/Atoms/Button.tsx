import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import type { RA } from '../../utils/types';
import { softFail } from '../Errors/Crash';
import { className } from './className';
import type { IconProps } from './Icons';
import { icons } from './Icons';
import { wrap } from './wrapper';

export const DialogContext = React.createContext<(() => void) | undefined>(
  undefined
);
DialogContext.displayName = 'DialogContext';

/**
 * A button that registers its onClick handler to containing dialog's
 * onClose handler.
 *
 * This is useful to avoid duplicating the dialog close logic
 * in the button's onClick and the dialog's onClose
 */
function DialogCloseButton({
  component: ButtonComponent = Button.Secondary,
  ...props
}: Omit<Parameters<typeof Button.Secondary>[0], 'onClick'> & {
  readonly component?: typeof Button.Secondary;
}): JSX.Element | null {
  const handleClose = React.useContext(DialogContext);
  if (handleClose === undefined) {
    softFail(new Error("Dialog's handleClose prop is undefined"));
    return null;
  }
  return <ButtonComponent {...props} onClick={handleClose} />;
}

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const button = (name: string, className: string) =>
  wrap<
    'button',
    {
      readonly onClick:
        | ((event: React.MouseEvent<HTMLButtonElement>) => void)
        | undefined;
      readonly children?:
        | JSX.Element
        | LocalizedString
        | RA<JSX.Element | LocalizedString | false | undefined>;
      readonly title?: LocalizedString | undefined;
      readonly 'aria-label'?: LocalizedString | undefined;
    }
  >(name, 'button', className, (props) => ({
    ...props,
    type: 'button',
    disabled: props.disabled === true || props.onClick === undefined,
  }));

export const Button = {
  /*
   * When using Button.LikeLink component, consider adding [role="link"] if the
   * element should be announced as a link
   */
  LikeLink: button('Button.LikeLink', className.link),
  Small: wrap<
    'button',
    {
      /*
       * A class name that is responsible for text and background color
       * Split into a separate prop in order to add a default value
       */
      readonly variant?: string;
      readonly onClick:
        | ((event: React.MouseEvent<HTMLButtonElement>) => void)
        | undefined;
    }
  >(
    'Button.Small',
    'button',
    className.smallButton,
    ({
      variant = className.defaultSmallButtonVariant,
      type,
      className: classString = '',
      disabled,
      ...props
    }) => ({
      type: 'button',
      className: `${classString} ${variant}`,
      disabled: disabled === true || props.onClick === undefined,
      ...props,
    })
  ),
  Fancy: button(
    'Button.LikeLink',
    `${className.niceButton} ${className.fancyButton}`
  ),
  Secondary: button(
    'Button.Secondary',
    `${className.niceButton} ${className.secondaryButton}`
  ),
  BorderedGray: button(
    'Button.BorderedGray',
    `${className.niceButton} ${className.borderedGrayButton}`
  ),
  Danger: button(
    'Button.Danger',
    `${className.niceButton} ${className.dangerButton}`
  ),
  Info: button(
    'Button.Info',
    `${className.niceButton} ${className.infoButton}`
  ),
  Warning: button(
    'Button.Warning',
    `${className.niceButton} ${className.warningButton}`
  ),
  Success: button(
    'Button.Success',
    `${className.niceButton} ${className.successButton}`
  ),
  Save: button(
    'Button.Save',
    `${className.niceButton} ${className.saveButton}`
  ),
  DialogClose: DialogCloseButton,
  Icon: wrap<
    'button',
    IconProps & {
      readonly onClick:
        | ((event: React.MouseEvent<HTMLButtonElement>) => void)
        | undefined;
    }
  >(
    'Button.Icon',
    'button',
    `${className.icon} rounded`,
    ({ icon, ...props }) => ({
      ...props,
      'aria-label': props['aria-label'] ?? props.title,
      type: 'button',
      disabled: props.disabled === true || props.onClick === undefined,
      children: icons[icon],
    })
  ),
} as const;
