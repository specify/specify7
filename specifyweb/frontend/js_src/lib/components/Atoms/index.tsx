/**
 * Primitive React components from which all other components are built
 *
 * These primitive components wrap native DOM elements, while also adding
 * custom styles and in some cases custom logic
 */

import { className } from './className';
import { wrap } from './wrapper';

export const ErrorMessage = wrap(
  'ErrorMessage',
  'div',
  'flex flex-col gap-2 p-2 text-white bg-red-500 rounded',
  {
    role: 'alert',
  }
);

export const Container = {
  /**
   * Full-screen gray container. Ment to be a wrapper for Container.Base
   */
  FullGray: wrap('Container.FullGray', 'div', className.containerFullGray),
  /**
   * Limited width white container. Ment to be wrapped inside Container.FullGray
   * Commonly used as an <aside> to main content
   */
  Base: wrap('Container.Base', 'section', className.containerBase),
  /**
   * Limited width white container. Ment to be wrapped inside Container.FullGray
   */
  Center: wrap('Container.Center', 'section', className.containerCenter),

  /**
   * Full-screen white container. Ment to be a wrapper for full width content
   */
  Full: wrap('Container.Full', 'section', className.containerFull),
};

export const Progress = wrap<'progress', { readonly value: number }>(
  'Progress',
  'progress',
  'w-full h-3 bg-gray-200 dark:bg-neutral-700 rounded',
  {
    max: 100,
  }
);

/*
 * Need to set explicit [role] for list without bullets to be announced as a list
 * REFACTOR: consider adding "flex flex-col gap-1" and .contents to <li>
 */
export const Ul = wrap('Ul', 'ul', '', { role: 'list' });

export const H2 = wrap('H2', 'h2', className.headerPrimary);
export const H3 = wrap('H3', 'h3', className.headerGray);

export const Summary = wrap<
  'summary',
  { readonly onToggle: (isCollapsed: boolean) => void }
>('Summary', 'summary', '', ({ onToggle: handleToggle, ...props }) => ({
  ...props,
  onClick:
    typeof props.onClick === 'function' || typeof handleToggle === 'function'
      ? (event): void => {
          /*
           * This is needed to prevent browser from handling state change
           * See: https://github.com/facebook/react/issues/15486
           */
          event.preventDefault();
          props.onClick?.(event);
          const details = (event.target as Element)?.closest('details');
          if (details === null)
            throw new Error("Can't use <summary> outside of <details>");
          handleToggle?.(!details.hasAttribute('open'));
        }
      : undefined,
}));

export const Key = wrap(
  'Key',
  'kbd',
  'bg-gray-200 border-1 dark:border-none dark:bg-neutral-700 rounded-sm mx-1 p-0.5'
);

export const oneRem = Number.parseFloat(
  getComputedStyle(document.documentElement).fontSize
);
