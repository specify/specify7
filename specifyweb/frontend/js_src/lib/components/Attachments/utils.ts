import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
