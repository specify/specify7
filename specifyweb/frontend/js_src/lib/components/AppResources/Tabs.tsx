import { Tab } from '@headlessui/react';
import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { resourcesText } from '../../localization/resources';
import { f } from '../../utils/functools';
import type { GetSet, IR, RA, RR } from '../../utils/types';
import { filterArray } from '../../utils/types';
import { WarningMessage } from '../Atoms';
import { toResource } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { getResourceApiUrl } from '../DataModel/resource';
import { schema } from '../DataModel/schema';
import type {
  SpAppResource,
  SpAppResourceDir,
  SpViewSetObj as SpViewSetObject,
} from '../DataModel/types';
import { ErrorBoundary } from '../Errors/ErrorBoundary';
import { Dialog, dialogClassNames } from '../Molecules/Dialog';
import { appResourceIcon } from './EditorComponents';
import { radioButtonClassName } from './Filters';
import { getAppResourceType, getResourceType } from './filtersHelpers';
import type {
  AppResourceEditorType,
  AppResourceTabProps,
} from './TabDefinitions';
import {
  AppResourceTextEditor,
  visualAppResourceEditors,
} from './TabDefinitions';

export function AppResourcesTab({
  tab: Component,
  label,
  headerButtons,
  appResource,
  resource,
  directory,
  data,
  isFullScreen: [isFullScreen, _],
  onChange: handleChange,
  onSetCleanup: setCleanup,
  footer,
}: {
  readonly tab: Component;
  readonly label: LocalizedString;
  readonly appResource: SpecifyResource<SpAppResource | SpViewSetObject>;
  readonly resource: SerializedResource<SpAppResource | SpViewSetObject>;
  readonly directory: SerializedResource<SpAppResourceDir>;
  readonly headerButtons: JSX.Element;
  readonly data: string | null;
  readonly isFullScreen: GetSet<boolean>;
  readonly onChange: (
    data: string | (() => string | null | undefined) | null
  ) => void;
  readonly onSetCleanup: (callback: () => Promise<void>) => void;
  readonly footer: JSX.Element;
}): JSX.Element {
  const children = (
    <ErrorBoundary dismissible>
      <Component
        appResource={appResource}
        data={data}
        directory={directory}
        resource={resource}
        onChange={handleChange}
        onSetCleanup={setCleanup}
      />
    </ErrorBoundary>
  );
  return isFullScreen ? (
    <Dialog
      buttons={<div className="flex gap-2">{footer}</div>}
      className={{
        container: dialogClassNames.fullScreen,
        buttonContainer: '!block',
      }}
      dimensionsKey={false}
      header={label}
      headerButtons={headerButtons}
      icon={appResourceIcon(getResourceType(resource))}
      onClose={undefined}
    >
      {children}
    </Dialog>
  ) : (
    children
  );
}

type Component = (props: AppResourceTabProps) => JSX.Element;

export function useEditorTabs(
  resource: SerializedResource<SpAppResource | SpViewSetObject>
): RA<{
  readonly label: LocalizedString;
  readonly component: (props: AppResourceTabProps) => JSX.Element;
}> {
  const subType =
    f.maybe(toResource(resource, 'SpAppResource'), getAppResourceType) ??
    'viewSet';
  return React.useMemo(() => {
    const editors =
      typeof subType === 'string'
        ? visualAppResourceEditors()[subType]
        : undefined;
    return editors === undefined
      ? [
          {
            label: labels.generic,
            component(props): JSX.Element {
              return <AppResourceTextEditor {...props} />;
            },
          },
        ]
      : filterArray(
          Object.entries(editors).map(([type, Editor]) =>
            typeof Editor === 'function'
              ? {
                  label: labels[type],
                  component(props): JSX.Element {
                    return (
                      <>
                        {type === 'visual' && (
                          <OtherCollectionWarning directory={props.directory} />
                        )}
                        <Editor {...props} />
                      </>
                    );
                  },
                }
              : undefined
          )
        );
  }, [subType]);
}

const labels: RR<AppResourceEditorType, LocalizedString> = {
  visual: resourcesText.visualEditor(),
  xml: resourcesText.xmlEditor(),
  json: resourcesText.jsonEditor(),
  generic: resourcesText.textEditor(),
};

/* Display a warning when editing resources from a different collection */
function OtherCollectionWarning({
  directory,
}: {
  readonly directory: SerializedResource<SpAppResourceDir>;
}): JSX.Element | null {
  const isOtherScope = React.useMemo(
    () =>
      (typeof directory.collection === 'string' &&
        directory.collection !==
          getResourceApiUrl('Collection', schema.domainLevelIds.collection)) ||
      (typeof directory.discipline === 'string' &&
        directory.discipline !==
          getResourceApiUrl('Discipline', schema.domainLevelIds.discipline)),
    [directory]
  );
  return isOtherScope ? (
    <WarningMessage>{resourcesText.wrongScopeWarning()}</WarningMessage>
  ) : null;
}

export function Tabs({
  tabs,
  index: [currentIndex, handleChange],
}: {
  readonly tabs: IR<JSX.Element>;
  readonly index: GetSet<number>;
}): JSX.Element {
  return (
    <Tab.Group selectedIndex={currentIndex} onChange={handleChange}>
      <Tab.List
        // Don't display tabs if there is only one tab
        className={`
          inline-flex w-fit flex-wrap gap-2 rounded
          bg-[color:var(--form-background)]
          ${Object.keys(tabs).length === 1 ? 'sr-only' : ''}
        `}
      >
        {Object.keys(tabs).map((label, index) => (
          <Tab
            className={radioButtonClassName(currentIndex === index)}
            key={index}
            /**
             * HeadlessUI does not trigger onChange on click on current tab.
             * This is a workaround. It overrides their click handler only
             * if the option IS current.
             */
            onClick={
              currentIndex === index
                ? (): void => handleChange(index)
                : undefined
            }
          >
            {label}
          </Tab>
        ))}
      </Tab.List>
      <Tab.Panels className="flex flex-1 overflow-hidden">
        {Object.values(tabs).map((element, index) => (
          <Tab.Panel className="flex flex-1 flex-col gap-4" key={index}>
            <ErrorBoundary dismissible>{element}</ErrorBoundary>
          </Tab.Panel>
        ))}
      </Tab.Panels>
    </Tab.Group>
  );
}
