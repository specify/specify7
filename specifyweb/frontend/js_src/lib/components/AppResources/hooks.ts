import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { useAsyncState } from '../../hooks/useAsyncState';
import { f } from '../../utils/functools';
import type { GetOrSet, IR, RA } from '../../utils/types';
import { addMissingFields } from '../DataModel/addMissingFields';
import { fetchCollection } from '../DataModel/collection';
import type { SerializedResource } from '../DataModel/helperTypes';
import type {
  Collection,
  Discipline,
  SpAppResource,
  SpAppResourceData,
  SpecifyUser,
  SpViewSetObj,
} from '../DataModel/types';
import { userPreferences } from '../Preferences/userPreferences';
import { getAppResourceType } from './filtersHelpers';
import { getAppResourceCount, getAppResourceMode } from './helpers';
import { getAppResourceTree, getScope } from './tree';
import type { ScopedAppResourceDir } from './types';
import { appResourceSubTypes } from './types';

export type AppResources = {
  readonly directories: RA<ScopedAppResourceDir>;
  readonly disciplines: RA<SerializedResource<Discipline>>;
  readonly collections: RA<SerializedResource<Collection>>;
  readonly users: RA<SerializedResource<SpecifyUser>>;
  readonly appResources: RA<SerializedResource<SpAppResource>>;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
};

export function useAppResources(
  loadingScreen: boolean = true
): GetOrSet<AppResources | undefined> {
  return useAsyncState(
    React.useCallback(
      async () =>
        f.all({
          directories: fetchCollection('SpAppResourceDir', {
            limit: 0,
            domainFilter: false,
          }).then<AppResources['directories']>(({ records }) =>
            records.map((record) => ({ ...record, scope: getScope(record) }))
          ),
          disciplines: fetchCollection('Discipline', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          collections: fetchCollection('Collection', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          users: fetchCollection('SpecifyUser', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          appResources: fetchCollection('SpAppResource', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          viewSets: fetchCollection('SpViewSetObj', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
        }),
      []
    ),
    loadingScreen
  );
}

export type AppResourcesTree = RA<{
  readonly label: LocalizedString;
  /*
   * A string that would be stable thought the lifecycle of an object.
   * Used to identify a tree node when storing conformation it in localStorage.
   *
   * Directory key is created instead of using id, because app resource
   * needs a way to identify directories that don't yet have IDs (i.e, when
   * creating an app resource for a collection that didn't have any app resources
   * before, the collection won't have any SpAppResourceDir record, thus not
   * directory ID).
   */
  readonly key: string;
  readonly directory: ScopedAppResourceDir | undefined;
  readonly appResources: RA<
    SerializedResource<SpAppResource> & {
      readonly label?: LocalizedString;
    }
  >;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
  readonly subCategories: AppResourcesTree;
}>;

export function useResourcesTree(resources: AppResources): AppResourcesTree {
  const [localize] = userPreferences.use(
    'appResources',
    'appearance',
    'localizeResourceNames'
  );
  return React.useMemo<AppResourcesTree>(() => {
    const tree = getAppResourceTree(resources);
    return localize ? localizeTree(tree) : tree;
  }, [resources, localize]);
}

const localizeTree = (tree: AppResourcesTree): AppResourcesTree =>
  tree.map(({ appResources, subCategories, ...rest }) => ({
    ...rest,
    appResources: appResources.map(localizeResource),
    subCategories: localizeTree(subCategories),
  }));

function localizeResource(
  resource: SerializedResource<SpAppResource> & {
    readonly label?: LocalizedString;
  }
): SerializedResource<SpAppResource> & { readonly label?: LocalizedString } {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  // Check that resource of this type can only have one specific name
  return typeof type.name === 'string'
    ? {
        ...resource,
        // Then replace the name with a localized label unless it's already set
        label: resource.label ?? type.label,
      }
    : resource;
}

export function useAppResourceCount(
  resourcesTree: AppResourcesTree[number]
): number {
  return React.useMemo(
    () => getAppResourceCount(resourcesTree),
    [resourcesTree]
  );
}

/**
 * Fetch resource contents
 */
export function useAppResourceData(
  resource: SerializedResource<SpAppResource | SpViewSetObj>,
  initialData: string | undefined
): {
  readonly resourceData: GetOrSet<
    SerializedResource<SpAppResourceData> | undefined
  >;
  readonly isChanged: boolean;
} {
  const initialValue = React.useRef<string>('');
  const [resourceData, setResourceData] = useAsyncState<
    SerializedResource<SpAppResourceData>
  >(
    React.useCallback(async () => {
      const relationshipName =
        getAppResourceMode(resource) === 'appResources'
          ? 'spAppResource'
          : 'spViewSetObj';
      const newResource = addMissingFields('SpAppResourceData', {
        [relationshipName]: resource.id,
        data: initialData,
      });
      const dataResource =
        typeof resource.id === 'number'
          ? await fetchCollection('SpAppResourceData', {
              limit: 1,
              [relationshipName]: resource.id,
              domainFilter: false,
            }).then(
              ({ records }) =>
                /*
                 * For some reason, app resource can have multiple app resource
                 * datas (but it never does in practice)
                 */
                records[0] ?? newResource
            )
          : newResource;
      const newData = fixLineBreaks(dataResource.data ?? '');
      initialValue.current = newData;
      return { ...dataResource, data: newData };
    }, [resource, initialData]),
    true
  );
  return {
    resourceData: [resourceData, setResourceData],
    isChanged: initialValue.current !== resourceData?.data,
  };
}

const fixLineBreaks = (string: string): string =>
  string.replaceAll(/[\n\r]+/gu, '\n');

export const getAppResourceExtension = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): string =>
  resource._tableName === 'SpViewSetObj'
    ? 'xml'
    : getResourceExtension(resource as SerializedResource<SpAppResource>);

function getResourceExtension(
  resource: SerializedResource<SpAppResource>
): 'jrxml' | 'json' | 'properties' | 'txt' | 'xml' {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  const mimeType = resource.mimeType?.toLowerCase() ?? type?.mimeType ?? '';
  if (mimeType in mimeMapper) return mimeMapper[mimeType];
  else if (mimeType.startsWith('jrxml')) return 'jrxml';
  else if (
    resource.name === 'preferences' &&
    mimeType === appResourceSubTypes.otherPropertiesResource.mimeType
  )
    return 'properties';
  else return 'txt';
}

const mimeMapper: IR<'json' | 'properties' | 'txt' | 'xml'> = {
  'application/json': 'json',
  'text/xml': 'xml',
  'text/plain': 'txt',
};
