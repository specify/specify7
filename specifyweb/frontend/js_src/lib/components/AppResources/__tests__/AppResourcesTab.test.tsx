import React from 'react';

import { clearIdStore } from '../../../hooks/useId';
import { requireContext } from '../../../tests/helpers';
import { mount } from '../../../tests/reactUtils';
import { f } from '../../../utils/functools';
import { localized } from '../../../utils/types';
import { Button } from '../../Atoms/Button';
import { deserializeResource } from '../../DataModel/serializers';
import { UnloadProtectsContext } from '../../Router/UnloadProtect';
import type { AppResourceTabProps } from '../TabDefinitions';
import { AppResourcesTab } from '../Tabs';
import { testAppResources } from './testAppResources';

beforeEach(() => {
  clearIdStore();
});

requireContext();

function Component(props: AppResourceTabProps) {
  return <h1>Data: {props.data}</h1>;
}

describe('AppResourcesTab', () => {
  test('simple render', () => {
    const { asFragment } = mount(
      <AppResourcesTab
        appResource={deserializeResource(testAppResources.appResources[0])}
        data="TestData"
        directory={testAppResources.directories[0]}
        footer={<h3>TestFooter</h3>}
        headerButtons={<Button.Info onClick={f.void} />}
        isFullScreen={[false, f.void]}
        label={localized('test')}
        resource={testAppResources.appResources[0]}
        tab={Component}
        onChange={f.void}
        onSetCleanup={f.void}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('dialog render', () => {
    const { getByRole } = mount(
      <UnloadProtectsContext.Provider value={[]}>
        <AppResourcesTab
          appResource={deserializeResource(testAppResources.appResources[0])}
          data="TestData"
          directory={testAppResources.directories[0]}
          footer={<h3>TestFooter</h3>}
          headerButtons={<Button.Info onClick={f.void} />}
          isFullScreen={[true, f.void]}
          label={localized('test')}
          resource={testAppResources.appResources[0]}
          tab={Component}
          onChange={f.void}
          onSetCleanup={f.void}
        />
      </UnloadProtectsContext.Provider>
    );

    const dialog = getByRole('dialog');
    expect(dialog).toMatchSnapshot();
  });
});
