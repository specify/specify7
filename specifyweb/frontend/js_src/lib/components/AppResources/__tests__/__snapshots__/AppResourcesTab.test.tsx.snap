// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppResourcesTab dialog render 1`] = `
<div
  aria-describedby="modal-0-content"
  aria-labelledby="modal-0-header"
  aria-modal="true"
  class="ReactModal__Content 
        flex flex-col gap-2 p-4 outline-none !transform-none !w-full !h-full
        overflow-x-hidden text-neutral-900 duration-0
        dark:border dark:border-neutral-700 dark:text-neutral-200
        
        
    bg-gradient-to-bl from-gray-200 via-white
    to-white dark:from-neutral-800 dark:via-neutral-900 dark:to-neutral-900
  
      "
  id="modal-0-full-screen"
  role="dialog"
  style="transform: translate(0px,0px);"
  tabindex="-1"
>
  <div
    class="
          flex items-center gap-2 md:gap-4
          
          flex-wrap
        "
    id="modal-0-handle"
  >
    <div
      class="flex items-center gap-2"
    >
      <span
        class="text-blue-500"
      >
        <span
          aria-hidden="true"
          title="Other Properties Resource"
        >
          <svg
            aria-hidden="true"
            class="w-6 h-6 flex-shrink-0"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M3 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zm0 4a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </span>
      <h2
        class="font-semibold text-black dark:text-white text-xl"
        id="modal-0-header"
      >
        test
      </h2>
    </div>
    <button
      class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-info hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--info-button-color)] text-white"
      type="button"
    />
  </div>
  <div
    class="
            dark:text-neutral-350 -mx-1 flex-1 overflow-y-auto px-1 py-4
            text-gray-700 flex flex-col gap-2
          "
    id="modal-0-content"
  >
    <h1>
      Data: 
      TestData
    </h1>
  </div>
  <div
    class="flex gap-2 !block"
  >
    <div
      class="flex gap-2"
    >
      <h3>
        TestFooter
      </h3>
    </div>
  </div>
</div>
`;

exports[`AppResourcesTab simple render 1`] = `
<DocumentFragment>
  <h1>
    Data: TestData
  </h1>
</DocumentFragment>
`;
